palette/moreland/luminance.go

```diff
@@ -0,0 +1,297 @@
+	// the possibility of the At() function returning
+	// an out-of-range error for values that actually
+	// should be in the range.
+	l.scalars[0] = 0
+	l.scalars[len(l.scalars)-1] = 1
+	return &l, nil
+}
+
+// At implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) At(v float64) (color.Color, error) {
+	if err := checkRange(l.min, l.max, v); err != nil {
+		return nil, err
+	}
+	scalar := (v - l.min) / l.max
+	if !inUnitRange(scalar) {
+		return nil, fmt.Errorf("moreland: interpolation value (%g) out of range [%g,%g]", scalar, l.min, l.max)
+	}
+	i := searchFloat64s(l.scalars, scalar)
+	if i == 0 {
+		return l.colors[i].cieXYZ().rgb().sRGBA(l.alpha), nil
+	}
+	c1 := l.colors[i-1]
+	c2 := l.colors[i]
+	frac := (scalar - l.scalars[i-1]) / (l.scalars[i] - l.scalars[i-1])
+	o := cieLAB{
+		L: frac*(c2.L-c1.L) + c1.L,
+		A: frac*(c2.A-c1.A) + c1.A,
+		B: frac*(c2.B-c1.B) + c1.B,
+	}.cieXYZ().rgb().sRGBA(l.alpha)
+	o.clamp()
+	return o, nil
+}
+
+func checkRange(min, max, val float64) error {
+	if max == min {
+		return fmt.Errorf("moreland: color map max == min == %g", max)
+	}
+	if min > max {
+		return fmt.Errorf("moreland: color map max (%g) < min (%g)", max, min)
+	}
+	if val < min {
+		return palette.ErrUnderflow
+	}
+	if val > max {
+		return palette.ErrOverflow
+	}
+	if math.IsNaN(val) {
+		return palette.ErrNaN
+	}
+	return nil
+}
+
+// searchFloat64s acts the same as sort.SearchFloat64s, except
+// it uses a simple search algorithm instead of binary search.
+func searchFloat64s(vals []float64, val float64) int {
+	for j, v := range vals {
+		if val <= v {
+			return j
+		}
+	}
+	return len(vals)
+}
+
+// SetMax implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) SetMax(v float64) {
+	l.max = v
+}
+
+// SetMin implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) SetMin(v float64) {
+	l.min = v
+}
+
+// Max implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) Max() float64 {
+	return l.max
+}
+
+// Min implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) Min() float64 {
+	return l.min
+}
+
+// SetAlpha sets the opacity value of this color map. Zero is transparent
+// and one is completely opaque.
+// The function will panic is alpha is not between zero and one.
+func (l *luminance) SetAlpha(alpha float64) {
+	if !inUnitRange(alpha) {
+		panic(fmt.Errorf("invalid alpha: %g", alpha))
+	}
+	l.alpha = alpha
+}
+
+// Alpha returns the opacity value of this color map.
+func (l *luminance) Alpha() float64 {
+	return l.alpha
+}
+
+// Palette returns a value that fulfills the palette.Palette interface,
+// where nColors is the number of desired colors.
+func (l luminance) Palette(n int) palette.Palette {
+	if l.max == 0 {
+		l.max = 1
+	}
+	delta := l.max / float64(n-1)
+	delta = math.Nextafter(delta, delta/2) // Subtract epsilon to avoid rounding problems.
```

Comment:
I still don't understand why this is necessary. If it truly is, `math.Nextafter(delta, 0)` does the same work without a division.